lib/srv/alpnproxy/conn.go

```diff
@@ -87,3 +91,106 @@ func (conn readOnlyConn) RemoteAddr() net.Addr               { return &utils.Net
 func (conn readOnlyConn) SetDeadline(t time.Time) error      { return nil }
 func (conn readOnlyConn) SetReadDeadline(t time.Time) error  { return nil }
 func (conn readOnlyConn) SetWriteDeadline(t time.Time) error { return nil }
+
+// newPingConn returns a ping connection wrapping the provided net.Conn.
+func newPingConn(conn net.Conn) *pingConn {
+	return &pingConn{Conn: conn}
+}
+
+// pingConn wraps a net.Conn and add ping capabilities to it, including the
+// `WritePing` function and `Read` (which excludes ping packets).
+//
+// When using this connection, the packets written will contain an initial data:
+// the packet size. When reading, this information is taken into account, but it
+// is not returned to the caller.
+//
+// Ping messages have a packet size of zero and are produced only when
+// `WritePing` is called. On `Read`, any Ping packet is discarded.
+type pingConn struct {
+	net.Conn
+
+	muRead  sync.Mutex
+	muWrite sync.Mutex
+
+	// currentSize size of bytes of the current packet.
+	currentSize uint32
+}
+
+// Read reads content from the underlaying connection, discarding any ping
+// messages it finds.
+func (c *pingConn) Read(p []byte) (int, error) {
+	c.muRead.Lock()
+	defer c.muRead.Unlock()
+
+	err := c.discardPingReads()
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+
+	// Check if the current size is larger than the provided buffer.
+	readSize := c.currentSize
+	if c.currentSize > uint32(len(p)) {
+		readSize = uint32(len(p))
+	}
+
+	n, err := c.Conn.Read(p[:readSize])
+	c.currentSize -= uint32(n)
+
+	return n, err
+}
+
+// WritePing writes the ping packet to the connection.
+func (c *pingConn) WritePing() error {
+	c.muWrite.Lock()
+	defer c.muWrite.Unlock()
+
+	return binary.Write(c.Conn, binary.BigEndian, uint32(0))
+}
+
+// discardPingReads reads from the wrapped net.Conn until it encounters a
+// non-ping packet.
+func (c *pingConn) discardPingReads() error {
+	for c.currentSize == 0 {
+		err := binary.Read(c.Conn, binary.BigEndian, &c.currentSize)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	return nil
+}
+
+// Write writes provided content to the underlaying connection with proper
+// protocol fields.
+func (c *pingConn) Write(p []byte) (int, error) {
+	c.muWrite.Lock()
+	defer c.muWrite.Unlock()
+
+	// Avoid casting into something bigger than acceptable.
+	if len(p) > math.MaxUint32 {
```

Comment:
Brief context: I've tried to break the connection using database dumps up to 5GB, but without success (their protocol handled it differently, usually with streaming). I've added this as a double-checker if any underlying protocol doesn't take it properly. It is also worth noting that this is a concern only with `Database -> Teleport` communications as it breaks the entire agent if not handled correctly.

With that being said, I think having a more reasonable limit will also handle this. What do you think, @greedy52 @smallinsky?

Reply:
How about having a `MaxWriteChunk` say default to ~20MB (MySQL default max packet is 16MB, Postgres is 20MB), and then if `len(p)` is greater than the chunk, instead of return an error, break `p` into chunks and then send each chunk in a loop?

Reply:
I would rather try to avoid this kind of logic in `pingConn` and try to keep things simple and relay on TCP MTU instead.

Reply:
For some database write protocols the max size of the package is quite large  for instance mysql says that max limit for single package is set to [1GB](https://dev.mysql.com/doc/refman/8.0/en/packet-too-large.html#:~:text=The%20largest%20possible%20packet%20that,server%20or%20client%20is%201GB.) 
> The largest possible packet that can be transmitted to or from a MySQL 8.0 server or client is 1GB.

Reply:
I see but this limitation should be enforced on db engine db proxy side where database packets are handled (this pingConn handler doesn't do any allocation). It is hard to anticipate what max packet limit size will be for database protocols there is always a risk that we will add a protocol that by default has a large limit like 64MB

Reply:
This check would only guarantee we're not panicking due to an invalid cast (which nothing would directly prevent from happening). I didn't consider it a problem since we're not copying data or doing any allocation (or at least it is not done on the `PingConn`), so limiting it here would not solve any potential issue.

If you feel like this is an unnecessary check, we can drop it. Still, I wouldn't give it the purpose of limiting what size can be sent. This limitation is particular with each database protocol we implement, so it seems better to deal with it per database protocol.

For curiosity, in my tests, I found that having those large packets would fail a step before, while generating the audit events:
```
2022-08-19T13:13:21-03:00 ERRO [DB:SERVIC] Audit write timed out after 5s. Will be losing events for the next 30s. events/auditwriter.go:361
2022-08-19T13:13:57-03:00 ERRO             Failed to emit audit event. error:[
ERROR REPORT:
Original Error: trace.aggregate grpc: received message larger than max (16777441 vs. 4194304)
Stack Trace:
        /teleport/lib/events/emitter.go:289 github.com/gravitational/teleport/lib/events.(*MultiEmitter).EmitAuditEvent
        /teleport/lib/events/emitter.go:168 github.com/gravitational/teleport/lib/events.(*CheckingEmitter).EmitAuditEvent
        /teleport/lib/events/emitter.go:94 github.com/gravitational/teleport/lib/events.(*AsyncEmitter).forward
        /opt/homebrew/Cellar/go/1.18.3/libexec/src/runtime/asm_arm64.s:1263 runtime.goexit
User Message: grpc: received message larger than max (16777441 vs. 4194304)] events/emitter.go:170
```

Reply:
@gabrielcorado Thanks for testing that. In this case, I'd leave it as it is now; just add a comment that we limit the package size at the GRPC level.